src/LinAlg/hiopMatrixMDS.hpp

```diff
@@ -332,29 +336,36 @@ class hiopMatrixSymBlockDiagMDS : public hiopMatrix
     assert(false && "not needed / implemented");
   }
 
-  virtual void addMatrix(double alpha, const hiopMatrix& X)
+  virtual void addMatrix(double alpha, const hiopMatrix& X_in)
   {
-    const hiopMatrixSymBlockDiagMDS* pX=dynamic_cast<const hiopMatrixSymBlockDiagMDS*>(&X);
-    if(pX==NULL) {
-      assert(false && "operation only supported for hiopMatrixMDS left operand");
-    } 
-    mSp->addMatrix(alpha, *pX->mSp);
-    mDe->addMatrix(alpha, *pX->mDe);
+    const hiopMatrixSymBlockDiagMDS& X = dynamic_cast<const hiopMatrixSymBlockDiagMDS&>(X_in);
+    mSp->addMatrix(alpha, *X.mSp);
+    mDe->addMatrix(alpha, *X.mDe);
   }
 
-  /* block of W += alpha*this */
+  /**
+   *  block of W += alpha*this
+   * 
+   * @todo Should this method be available at all? Consider using
+   * addUpperTriangleToSymDenseMatrixUpperTriangle instead.
+   */
   virtual void addToSymDenseMatrixUpperTriangle(int row_start, int col_start, double alpha, hiopMatrixDense& W) const
   {
     assert(mSp->m() == mSp->n());
     mSp->addToSymDenseMatrixUpperTriangle(row_start,          col_start,          alpha, W);
     mDe->addToSymDenseMatrixUpperTriangle(row_start+mSp->n(), col_start+mSp->n(), alpha, W);
+    assert(0 && "This should not be called for symmetric matrices.");
   }
-  /* block of W += alpha*this */
+  /**
+   *  block of W += alpha*this
+   * 
+   * @warning This method is implemented correctly, but should never be called actually.
+   * Use addUpperTriangleToSymDenseMatrixUpperTriangle instead.
+   */
   virtual void transAddToSymDenseMatrixUpperTriangle(int row_start, int col_start, double alpha, hiopMatrixDense& W) const
   {
-    assert(mSp->m() == mSp->n());
-    mSp->transAddToSymDenseMatrixUpperTriangle(row_start,          col_start,          alpha, W);
-    mDe->transAddToSymDenseMatrixUpperTriangle(row_start+mSp->n(), col_start+mSp->n(), alpha, W);
+    addToSymDenseMatrixUpperTriangle(row_start, col_start, alpha, W);
```

Comment:
I'm questioning the decision to place this function call above the assertion. If some user were to call this function (which we don't want them to do) and an error is raised deeper into the call to `addToSymDenseMatrixUpperTriangle`, the error would be very misleading and potentially hard to trace down. @pelesh why did we leave function calls like this one in the code when we will just end up asserting afterwards?

Reply:
see https://github.com/LLNL/hiop/issues/98

`addToSymDenseMatrixUpperTriangle` and `transAddToSymDenseMatrixUpperTriangle` were replaced with `addUpperTriangleToSymDenseMatrixUpperTriangle`

Reply:
see a4022c8 and comments in #98

Reply:
The unit test for `addToSymDenseMatrixUpperTriangle` in symmetric sparse matrix class  should be kept to test the symmetric sparse matrix method `addUpperTriangleToSymDenseMatrixUpperTriangle`. It unit test was already doing that, because `addUpperTriangleToSymDenseMatrixUpperTriangle` was implemented as a wrapper to `addToSymDenseMatrixUpperTriangle` with row offset set to be equal to the column offset. We implemented test for the wrapped function instead of the wrapper, but simply replacing the names in the unit test would make it work again.

Reply:
will reactivate test for `addToSymDenseMatrixUpperTriangle` as test for `addUpperTriangleToSymDenseMatrixUpperTriangle`

Reply:
> will reactivate test for addToSymDenseMatrixUpperTriangle as test for addUpperTriangleToSymDenseMatrixUpperTriangle

I've got it. Will push shortly.

Reply:
Fixed in 3d15a97.